### PR TITLE
Support MiMo v2 ASR

### PR DESCRIPTION
**Pull request by zalcit**
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #27058738757](https://github.com/sgl-project/sglang/actions/runs/27058738757)
Latest PR Test (Extra): :x: [Run #27058738710](https://github.com/sgl-project/sglang/actions/runs/27058738710)

### CODE REVIEW

File: python/sglang/srt/models/mimo_audio.py [hunk 1]
```diff
@@ -1141,88 +1142,118 @@ def output_local_config(self):
...
-        model_path = self.server_args.model_path
-        if not os.path.isdir(model_path):
-            from huggingface_hub import snapshot_download
+            raise ValueError(f"Invalid projection layers: {config.projection_layers}")
+
+        # Precedence: --audio-tokenizer-path CLI flag > config.audio_tokenizer_path
+        # > <model_path>/audio_tokenizer fallback. The CLI flag lets users point at
+        # a relocated tokenizer dir without editing the model's config.json.
+        server_args = get_global_server_args()
+        cli_audio_tokenizer_path = getattr(server_args, "audio_tokenizer_path", None)
```
[the remark this answers is not in the document]
Comment (zalcit) [reply python/sglang/srt/models/mimo_audio.py]: get_global_server_args() raises rather than returns None — see server_args.py:7599. The getattr(..., None) already handles the case where audio_tokenizer_path attr is
   missing, which is the only realistic failure mode here.

File: python/sglang/srt/models/mimo_v2_asr.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    except ImportError as e:
+        raise RuntimeError(
+            "MiMo-V2-ASR audio encoder needs upstream flash-attn on this GPU "
+            "(sgl-kernel FA3 doesn't support sm_120). Install with "
+            "`pip install flash-attn --no-build-isolation`."
+        ) from e
+    _mimo_audio_module.flash_attn_varlen_func = flash_attn_varlen_func
+
+
+MiMoV2ASRConfig = None
```
[the remark this answers is not in the document]
Comment (zalcit) [reply python/sglang/srt/models/mimo_v2_asr.py]: Agreed on switching to Any to drop the # type: ignore. The runtime concern doesn't apply though — annotations aren't evaluated at call time, so
  config.audio_config was never affected by the placeholder being None.

File: python/sglang/srt/multimodal/processors/mimo_audio.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
+"""Stateful audio preprocessing pipeline shared by MiMo multimodal and ASR processors."""
+
+import io
+import math
+import os
```
Comment (ShangmingCai) [inline python/sglang/srt/multimodal/processors/mimo_audio.py:5]: Need to remove this to fix lint

**Comment by mickqian** [comment]
cc @AgainstEntropy

File: python/sglang/srt/multimodal/processors/mimo_v2.py [hunk 4]
```diff
@@ -704,7 +617,7 @@ def preprocess_for_encoder(self, mm_data, modality):
             for audio in mm_data:
                 if isinstance(audio, np.ndarray):
                     audio = (torch.from_numpy(audio).float(), self.audio_sampling_rate)
-                spec, token_len = self.preprocess_audio(audio)
+                spec, token_len = self.audio_pipeline.preprocess_audio(audio)
```
Comment (ShangmingCai) [inline python/sglang/srt/multimodal/processors/mimo_v2.py:626]: Should this be `self.audio_pipeline.audio_sampling_rate`?

File: python/sglang/srt/multimodal/processors/mimo_v2.py [hunk 5]
```diff
@@ -411,60 +359,24 @@ def __init__(
...
+            audio_n_mels=audio_n_mels,
+            audio_input_id_per_second=audio_input_id_per_second,
         )
-        self._mel_spectrogram = None
-        self._resamplers = OrderedDict()
-        self._resamplers_max = 16
-
-        self.audio_channels = audio_channels
-        self.audio_group_size = audio_group_size
-        self.audio_input_id_per_second = audio_input_id_per_second
```
Comment (ShangmingCai) [inline python/sglang/srt/multimodal/processors/mimo_v2.py:445]: Maybe keep these two as well? Or remove `audio_token_per_second` from MiMoProcessor. If not removed, you might need to check and replace the impl with `self.audio_pipeline.xxx`.

Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2.py:445]: Ok. They are the same issue. I will go through and check all the properties.

Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2.py:445]: Changing everything to self.audio_pipeline.xxx makes the code a bit bloated, so I’m planning to implement it using a delegate class instead.

File: python/sglang/srt/multimodal/processors/mimo_v2.py
Comment (ShangmingCai) [inline python/sglang/srt/multimodal/processors/mimo_v2.py]: Maybe need to use `self.audio_pipeline.process_audio` here as well?

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                if modality == Modality.AUDIO:
+                    loaded_audio = next(loaded_audio_iter)
+                    raw_audio_item = next(raw_audio_iter)
+
+                    if isinstance(loaded_audio, np.ndarray):
+                        audio_source = loaded_audio
+                    elif isinstance(raw_audio_item, dict):
+                        audio_source = raw_audio_item.get("url", loaded_audio)
+                    elif isinstance(raw_audio_item, (str, bytes, torch.Tensor)):
+                        audio_source = raw_audio_item
```
Comment (ShangmingCai) [inline python/sglang/srt/multimodal/processors/mimo_v2_asr.py:234]: Maybe add an else branch, and throw an error or log a warning here?

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+                if isinstance(content.content, str):
+                    input_ids.extend(self.tokenizer.encode(content.content))
+                else:
+                    input_ids.extend(content.content)
+            elif content.type == "audio":
+                result = self.audio_pipeline.process_audio_input(content.content)
+                audio_inputs.append(result["audio_input"])
+                input_ids.extend(result["input_ids"])
+
+        ids = torch.tensor(input_ids)
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/mimo_v2_asr.py]: `torch.as_tensor`?

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        position_ids = torch.arange(ids.shape[0]).expand(3, -1)
+        rope_deltas = torch.zeros((1, 1), dtype=torch.int32)
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/mimo_v2_asr.py:114]: nit: this could be cached

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        image_data: List[Union[str, bytes]],
+        audio_data: List[Union[str, bytes]],
+        input_text,
+        request_obj,
+        *args,
+        **kwargs,
+    ):
+        if audio_data is None:
+            audio_data = getattr(request_obj, "audio_data", [])
+        if audio_data and not self.AUDIO_REGEX.search(input_text):
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/mimo_v2_asr.py]: early return when `audio_data` is none here?

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+                    self.io_executor,
+                    lambda: self._process_contents(contents),
+                )
+            )
+        except RuntimeError as e:
+            logger.error(f"MiMo ASR processor failed in process_mm_data_async: {e}")
+            raise ValueError(f"Multimodal data is corrupted or cannot be decoded: {e}")
+
+        input_ids_flat = input_ids.flatten()
+        mm_items: List[MultimodalDataItem] = []
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/mimo_v2_asr.py]:
```python
if audio_inputs:
   mm_items = [...]
else:
   mm_items = []
```

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py (hunk 7 above)
Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2_asr.py]: Done, switched to

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py (hunk 8 above)
Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2_asr.py:114]: Skipping for consistency with the same pattern in mimo_v2.py / qwen3_asr.py — can unify in a follow-up.

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py (hunk 9 above)
Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2_asr.py]: Good point, added.

File: python/sglang/srt/multimodal/processors/mimo_v2_asr.py (hunk 10 above)
Comment (zalcit) [reply python/sglang/srt/multimodal/processors/mimo_v2_asr.py]: Done. Note: with the new early-return on empty audio_data, the else branch is effectively unreachable — happy to drop it if you'd prefer.

**Comment by zalcit** [comment]
@AgainstEntropy Can you check whether this PR is ready for merge? :D

File: python/sglang/srt/server_args.py [hunk 11]
```diff
@@ -358,6 +355,7 @@ class ServerArgs:
     tokenizer_worker_num: int = 1
     detokenizer_worker_num: int = 1
     skip_tokenizer_init: bool = False
+    audio_tokenizer_path: Optional[str] = None
```
Comment (ShangmingCai) [inline python/sglang/srt/server_args.py]: Is it possible to auto-detect and load the audio tokenizer in `mimo_audio.py`? Or get it by post-processing `model_path` and `tokenizer_path`? If this is achievable, we should avoid adding a new server arg.

Comment (zalcit) [reply python/sglang/srt/server_args.py]: Of course.

**Review by ShangmingCai — APPROVED** [review]

**Comment by AgainstEntropy** [comment]
hi @zalcit thank you for the model support!
Since it's an ASR model, can you also enable the `/v1/audio/transcriptions` api by simply implementing a `MiMoV2ASRAdapter`?

**Comment by zalcit** [comment]
> hi @zalcit thank you for the model support! Since it's an ASR model, can you also enable the `/v1/audio/transcriptions` api by simply implementing a `MiMoV2ASRAdapter`?

@AgainstEntropy done this

**Review by AgainstEntropy — APPROVED** [review]

**Review by yuan-luo — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
